sdk/storage/src/datalake/file_system_client.cpp

```diff
@@ -0,0 +1,225 @@
+    }
+    else
+    {
+      throw std::runtime_error("invalid connection string");
+    }
+
+    builder.AppendPath(fileSystemName, true);
+    auto credential = std::make_shared<SharedKeyCredential>(accountName, accountKey);
+
+    return FileSystemClient(builder.to_string(), credential, options);
+  }
+
+  FileSystemClient::FileSystemClient(
+      const std::string& fileSystemUri,
+      std::shared_ptr<SharedKeyCredential> credential,
+      const FileSystemClientOptions& options)
+      : m_fileSystemUri(fileSystemUri)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> policies;
+    for (const auto& p : options.policies)
+    {
+      policies.emplace_back(std::unique_ptr<Azure::Core::Http::HttpPolicy>(p->Clone()));
+    }
+    policies.emplace_back(std::make_unique<CommonHeadersRequestPolicy>());
+    policies.emplace_back(std::make_unique<SharedKeyPolicy>(credential));
+    policies.emplace_back(std::make_unique<Azure::Core::Http::TransportPolicy>(
+        std::make_shared<Azure::Core::Http::CurlTransport>()));
+    m_pipeline = std::make_shared<Azure::Core::Http::HttpPipeline>(policies);
+  }
+
+  FileSystemClient::FileSystemClient(
+      const std::string& fileSystemUri,
+      std::shared_ptr<TokenCredential> credential,
+      const FileSystemClientOptions& options)
+      : m_fileSystemUri(fileSystemUri)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> policies;
+    for (const auto& p : options.policies)
+    {
+      policies.emplace_back(std::unique_ptr<Azure::Core::Http::HttpPolicy>(p->Clone()));
+    }
+    policies.emplace_back(std::make_unique<CommonHeadersRequestPolicy>());
+    // not implemented yet
+    unused(credential);
+    policies.emplace_back(std::make_unique<Azure::Core::Http::TransportPolicy>(
+        std::make_shared<Azure::Core::Http::CurlTransport>()));
+    m_pipeline = std::make_shared<Azure::Core::Http::HttpPipeline>(policies);
+  }
+
+  FileSystemClient::FileSystemClient(
+      const std::string& fileSystemUri,
+      const FileSystemClientOptions& options)
+      : m_fileSystemUri(fileSystemUri)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> policies;
+    for (const auto& p : options.policies)
+    {
+      policies.emplace_back(std::unique_ptr<Azure::Core::Http::HttpPolicy>(p->Clone()));
+    }
+    policies.emplace_back(std::make_unique<CommonHeadersRequestPolicy>());
+    policies.emplace_back(std::make_unique<Azure::Core::Http::TransportPolicy>(
+        std::make_shared<Azure::Core::Http::CurlTransport>()));
+    m_pipeline = std::make_shared<Azure::Core::Http::HttpPipeline>(policies);
+  }
+
+  PathClient FileSystemClient::GetPathClient(const std::string& path) const
+  {
+    PathClient client = PathClient();
+    auto builder = m_fileSystemUri; // Copy
+    builder.AppendPath(path, true);
+    client.m_pathUri = std::move(builder);
+    client.m_pipeline = m_pipeline; // Copy
```

Comment:
ditto

Reply:
It's like we copied a shared pointer. big deal. we don't need to make everybody know that.

Reply:
Removed.